src/map/npc.c

```diff
@@ -3026,7 +3026,7 @@ static int npc_unload(struct npc_data *nd, bool single, bool unload_mobs)
 {
 	nullpo_ret(nd);
 
-	if (nd->ud != NULL && nd->ud != &npc->base_ud)
+	if (nd->ud != NULL)
```

Comment:
this mean if unload one npc, it will remove units from all npcs?

Reply:
probably need call unit->bl2ud2 if adding any unit group (call in skill_initunitgroup)?

Reply:
>this mean if unload one npc, it will remove units from all npcs?
No, only from the one that was passed to `npc_unload()`.

> probably need call unit->bl2ud2 if adding any unit group (call in skill_initunitgroup)?
This would fix the bug, too.

Reply:
if ud == npc->base_ud, this mean unit_data struct is shared between all default looks npcs. and unit_groups look like part or unit_data. this mean if change in base ud, it will affect any npc with same ud. npc id here completely unused

Reply:
In `skill_initunitgroup()` the NPC's ID is assigned like this:
`group->src_id      = src->id;` (Where `src` is the NPC's `block_list`.)

Reply:
While adding a call to `unit->bl2ud2()` in `skill_initunitgroup()` might be worse for performance, it is probably the cleanest approach, since it'd also solve a problem that I can expect to happen right now: if more than `MAX_SKILLUNITGROUP` NPCs in different locations use `unitskillusepos()` in rapid succession, and they end up deleting each others' unit groups because the `ud->skillunit[]` is full.

Reply:
Changed. :smile: